Added location to user defined attributes

AsterMiha opened the pull request:
Created a new node (UadItem) to preserve the location of all user defined attributes

Commit:
Added location to user defined attributes

ibuclaw commented on src/dmd/parse.d:
```diff
@@ -1445,25 +1446,29 @@ final class Parser(AST) : Lexer
                 const loc = token.loc;
                 exp = new AST.CallExp(loc, exp, parseArguments());
+                parens = false;
             }
 
             if (udas is null)
                 udas = new AST.Expressions();
-            udas.push(exp);
+            udas.push(new AST.UadItem(udastart, exp, parens));
```
And why can't `exp` have its loc set to udastart?

AsterMiha replied:
Because udas can be written in any of these ways: @exp, @(exp), @(exp1, exp2)
In the first 2 cases changing the loc of the exp would work but in the last it would be confusing to set all of them to the loc of  '@', especially when the expressions are on different lines.

Geod24 replied:
I also think that introducing a new AST type to represent this is overkill, and will not scale well.

TungstenHeart reviewed (changes requested):
That's a lot of changes to have the position so I wont give my mind on this aspect. 
The changes requested are more about

1. The coverage that is not really good. 
2. The style a bit inconsistent with the use of `isUadItem()` and most importantly the the name of the new node is not good. the offical name for theose attribs are UDA, so `UDAItem` is better.
3. ASTBase declaration

TungstenHeart commented on src/dmd/semantic2.d:
```diff
@@ -570,13 +573,19 @@ private extern(C++) final class Semantic2Visitor : Visitor
                             TupleExp te = cast(TupleExp)e;
                             eval(sc, te.exps, lastTag);
                         }
+                        if (e.op == TOK.at)
```
However now I see that the way you write the cast is consistent with the other cast in the block.

MoonlightSentinel commented on src/dmd/expression.d:
```diff
@@ -282,6 +282,17 @@ extern (C++) void expandTuples(Expressions* exps)
         }
+
+        // Inline expand all the uadItems
+        while (arg.op == TOK.at)
```
No, DMD supports bootstrapping from 2.079

ibuclaw replied:
2.076 if you include GDC.

jacob-carlborg commented:
I will not do an extended review of this, but the tests are a bit lacking. They need to include:

* UDAs on multiple lines, i.e:
    ```d
    @(3)
    @(4)
    int a;
    ```
* Both of the other two UDA syntaxes (if those are indented to be supported by this PR), i.e.:
    ```d
    @(3) {
        int a;
    }
    ```
    ```d
    @(3):
        int a;
    ```

AsterMiha commented:
@TungstenHeart Renamed the node and added the ASTBase declaration. I forgot to remove some of the code from previous attempts and that caused the coverage issues, but I fixed that now.

TungstenHeart commented:
thanks, (someone can cancel my request for changes)

AsterMiha commented on src/dmd/dtemplate.d:
```diff
@@ -7575,7 +7575,7 @@ void unSpeculative(Scope* sc, RootObject o)
 {
     // None of these can be value parameters
-    if (e.op == TOK.tuple || e.op == TOK.scope_ ||
+    if (e.op == TOK.tuple || e.op == TOK.at || e.op == TOK.scope_ ||
```
Semantic2 attempts to interpret all expressions in UserAttributeDeclaration [1]. Now that the actual expressions are inside a UDAItem node we need to reach its contents first. An alternative would be to add UDAItem to the Interpretor visitor to unpack the node there.

[1] https://github.com/dlang/dmd/blob/2453350ae801d689d25f2b89d3a63897091fdfbb/src/dmd/semantic2.d#L573-L574

RazvanN7 reviewed (approved):
Looks good to me. I'll wait for @jacob-carlborg 's review

jacob-carlborg commented:
> I'll wait for jacob-carlborg 's review

I've already done the review I'll be doing.

TungstenHeart commented:
SInce this was not discussed (here at least), what is the value added , what are the use cases ?

RazvanN7 commented:
Currently, UDAs do not contain the location, as a third-party tools using dmd-as-a-lib that want to properly expose the location of UDAs cannot do that. This PR fixes that. More info: https://github.com/dlang/dmd/pull/11788#issuecomment-698186023

TungstenHeart commented on src/dmd/expressionsem.d:
```diff
@@ -3103,6 +3103,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+        {
+            Expression exp = (*e.exps)[i];
+            exp = exp.expressionSemantic(sc);
+            if (!exp.type)
```
`assert(exp.type)` in case you dont manage to reproduce the case.

TungstenHeart replied:
to be clear. If at some point someone one day compiles code that triggers the assert failure that will allow to spot a place were expsema returns null instead of errorexp.

TungstenHeart commented on src/dmd/expressionsem.d:
```diff
@@ -3103,6 +3103,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+                err = true;
+            }
+            else if (exp.op == TOK.error)
+                err = true;
```
could `return setError()` here

TungstenHeart reviewed (changes requested):
see comments. Mostly the assert that will avoid to merge possibly dead code.
